### PR TITLE
[FIX] payment: Exclude acquirers requiring tokenization when logged out

### PR DESCRIPTION
**Pull request by robinengels**
Current behavior:
Public user have access to payment method that require tokenization wich leads to error when they use it.
We should not show those payment methos to the public users (e.g We shouldn't show sepa direct debit because this method uses tokenization)

Steps to reproduce:
- Activate SEPA Direct Debit in the payment acquierers
- Select Bank as payment journal
- In the bank payment journal create an account number (e.g BE71096123456769), and select a Bank (e.g BNP)
- Create a new customer (type : company, country : spain, and an email)
- Create a quotation with this new customer
- Click on the "action" icon and click on generate payment link
- Open the link in an incognito window (to make sure you'r not connected in Odoo)
- Try to use the sepa direct debit
- An access error appears because public user do not have access to payment.token.

The error is happening here when trying to access the token 

https://github.com/odoo/odoo/blob/65cb9ef253f803c61a702569b06af6bf2d16b2cf/addons/payment/controllers/portal.py#L258-L262

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by agr-odoo** [comment]
Not sure about this:
- `sdd_mandate_id` is not in `payment` module so this would fail in the `except` case whenever `payment_sepa_direct_debit` is not installed
- Then if we don't want to create/use a mandate for public users, why do we create one and then delete when the transaction fails? 

Imho it would be better to add a check when creating the mandate
https://github.com/odoo/enterprise/blob/0a85eb01992a7fc9d1a7586c58042b1ece4441cc/payment_sepa_direct_debit/controllers/main.py#L87
something like
```python
if (request.uid == request.env.ref('base.public_user').id):
    raise UserError(_("You need to be logged in to use this payment method."))
```
Maybe @arj-odoo could give us useful info too

**Comment by arj-odoo** [comment]
Hello. I can confirm that you can't use sdd mandate feature in `payment` module. You must make your change in sepa_direct_debit associated modules. (e.g. payment_sepa_direct_debit).
You could either restrict the mandate creation to portal/internal users or create a partner on the fly and assign the mandate to him. I let @AntoineVDV answer as he is in charge of the payment module.

**Comment by AntoineVDV** [comment]
Hello, could you clarify the problem description a bit? I had to dig into the ticket and the attached videos to figure the issue out. :sweat_smile: 

So, the problem seems to be: a customer tries to pay something with SEPA while being logged out, an AccessError is raised somewhere along the way, and the mandate (maybe also the token) is still saved. I guess that the error is raised as soon as we try to read on the token [here](https://github.com/odoo/odoo/blob/15.0/addons/payment/controllers/portal.py#L260) without sudo, after creating it in sudo.

IMO, we shouldn't try to delete the mandate if something fails, and certainly not from `payment`'s code as it violates the modularity. Instead, we should prevent SEPA from being made available to the public user altogether. This could typically be done through an override of `_get_compatible_acquirers` on the `payment.acquirer` model. But, even though the use-case is currently specific to SEPA, I believe we should work in the `payment` module to fix the issue and avoid running into it again with new payment modules exhibiting the same use-case. The easiest way to achieve this is to add a new filter making use of the `_is_tokenization_required` method in `_get_compatible_acquirers`:

```py
compatible_acquirers = self.env['payment.acquirer'].search(domain)

# Prevent the public user from saving a token for acquirers that require tokenization.
if request.env.user._is_public():
    compatible_acquirers = compatible_acquirers.filtered(
        lambda acq: not self._is_tokenization_required(provider=acq.provider)
    )

return compatible_acquirers
```

File: addons/payment/models/payment_acquirer.py [hunk 1]
```diff
@@ -314,6 +314,12 @@ def _get_compatible_acquirers(
             domain = expression.AND([domain, [('allow_tokenization', '=', True)]])
 
         compatible_acquirers = self.env['payment.acquirer'].search(domain)
+
+        if self.env.user._is_public():
+            compatible_acquirers = compatible_acquirers.filtered(
+                lambda acq: not self._is_tokenization_required(provider=acq.provider)
```
Comment (AntoineVDV) [inline addons/payment/models/payment_acquirer.py:320]:
```suggestion
                lambda acq: not self._is_tokenization_required(provider=acq.provider, **kwargs)
```
Should not be necessary but better to be safe than sorry.

File: addons/payment/models/payment_acquirer.py [hunk 2]
```diff
@@ -314,6 +314,12 @@ def _get_compatible_acquirers(
             domain = expression.AND([domain, [('allow_tokenization', '=', True)]])
 
         compatible_acquirers = self.env['payment.acquirer'].search(domain)
+
+        if self.env.user._is_public():
```
Comment (AntoineVDV) [inline addons/payment/models/payment_acquirer.py:318]:
```suggestion
        # Prevent the public user from saving a token for acquirers that require tokenization.
        if self.env.user._is_public():
```

**Review by AntoineVDV — COMMENTED** [review]
Don't forget to:
1. Flag your PR as "ready for review" when the Bugfix team considers it ready.
2. Re-request a review with the :recycle: button to let reviewers know that you're ready for a new pass of review.

**Review by AntoineVDV — APPROVED** [review]
Don't forget to update both the commit title+message and the PR title+description to:
1. Make the issue clear: how to reproduce; what is the error; etc.
2. State what your commit is changing to the code. For example, `exclude acquirers requiring tokenization when logged out`.

**Review by AntoineVDV — CHANGES_REQUESTED** [review]
Please find in bold below the unresolved requests. In particular, the SEPA Direct Debit payment option is not a payment method, it's an acquirer. It's confusing to mix the two.

- Make the issue clear: how to reproduce; **what is the error**; etc.
- State what your commit is changing to the code. For example, "**[FIX] payment: exclude acquirers requiring tokenization when logged out**"

**Comment by robinengels** [comment]
@AntoineVDV This should be good now, sorry if I didn't understand what you wanted me to do exactly

**Review by AntoineVDV — APPROVED** [review]
:+1:

**Pull request closed without merging**